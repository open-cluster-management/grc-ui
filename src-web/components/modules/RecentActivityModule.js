/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { SECURITY_TYPES } from '../../../lib/shared/constants'
import resources from '../../../lib/shared/resources'
import msgs from '../../../nls/platform.properties'
import TopInformationModule from './TopInformationModule'
import * as d3 from 'd3'
import _ from 'lodash'

const arcGenerator = d3.arc()

resources(() => {
  require('../../../scss/module-recent-activity.scss')
})

const $grc_color_finding_high = '#E62325'
const $grc_color_finding_medium = '#FEC233'
const $grc_color_finding_low = '#A0A0A0'
const $grc_color_finding_ring = '#DFE3E6'

const MAX_INFORMATION_THRESHOLD = 4

export default class RecentActivityModule extends React.Component {
  constructor(props) {
    super(props)
    this.state = {
      threshold: 4,
      topViolationsNum: 0,
      topFindingsNum: 0,
    }
  }

  setThreshold = (numFindings, numViolations) => {
    //if max(Top violations, Top security findings) < 4, set information_threshold to max
    const newMax = Math.max(numFindings, numViolations)
    if (newMax > MAX_INFORMATION_THRESHOLD) {
      return MAX_INFORMATION_THRESHOLD
    }
    return newMax
  }

  thresholdCallback = (cardsLength, cardType) => {
    //updates module lengths and changes threshold if needed
    if (cardType === 'policies' && cardsLength !== this.state.topViolationsNum) {
      this.setState(
        { topViolationsNum: cardsLength },
        () => {
          const findingsN = this.state.topFindingsNum
          const violationsN = this.state.topViolationsNum
          this.setState(
            // need at least one threshold for top violations empty card
            { threshold: Math.max(1, this.setThreshold(findingsN, violationsN)) },
            () => {
              return
            }
          )
        }
      )
    }
    else if (cardType === 'findings' && cardsLength !== this.state.topFindingsNum) {
      this.setState(
        { topFindingsNum: cardsLength },
        () => {
          const findingsN = this.state.topFindingsNum
          const violationsN = this.state.topViolationsNum
          this.setState(
            // need at least one threshold for top finding empty card
            { threshold: Math.max(1, this.setThreshold(findingsN, violationsN)) },
            () => {
              return
            }
          )
        }
      )
    }
  }

  render() {
    const { locale } = this.context
    const title = msgs.get('overview.recent.activity.title', locale)
    const { handleDrillDownClick, viewState, policies, showFindings, findings, applications, updateViewState, showApplications } = this.props
    const moduleData = this.getModuleData()
    return (
      <div className='module-recent-activity'>
        <div className='card-container-container'>
          <div className='card-title'>
            {title}
          </div>
          <div className='card-container'>
            <div className='card-content'>
              <div className='card-inner-content'>
                <Violations moduleData={moduleData} handleDrillDownClick={handleDrillDownClick} locale={locale} />
                {showFindings ? <Findings moduleData={moduleData} handleDrillDownClick={handleDrillDownClick} locale={locale} /> : null}
              </div>
            </div>
          </div>
          <div className='violation-container'>
            <TopInformationModule
              type={'policies'}
              viewState={viewState.topViolationsChoice ? viewState.topViolationsChoice : 'policies'}
              updateViewState={updateViewState}
              items={policies}
              applications={applications}
              threshold={this.state.threshold}
              updateThreshold={this.thresholdCallback}
              handleDrillDownClick={handleDrillDownClick}
              showApplications={showApplications}
            />
            {showFindings ? <TopInformationModule
              type={'findings'}
              viewState={viewState.topFindingChoice ? viewState.topFindingChoice : 'findings'}
              updateViewState={updateViewState}
              items={findings}
              threshold={this.state.threshold}
              updateThreshold={this.thresholdCallback}
              handleDrillDownClick={handleDrillDownClick}
            /> : null}
          </div>
        </div>
      </div>
    )
  }

  getModuleData = () => {
    const { locale } = this.context
    const { policies, findings } = this.props
    const policySet = new Set()
    const clusterSet = new Set()
    policies.map(policy=>{
      const statuses = _.get(policy, 'raw.status.status', {})
      Object.keys(statuses).forEach(key=>{
        const compliant = statuses[key].compliant
        if (!compliant || compliant.toLowerCase()==='noncompliant') {
          clusterSet.add(key)
          policySet.add(_.get(policy, 'metadata.name', 'unknown'))
        }
      })
    })
    let low = 0, medium = 0, high = 0
    const total = findings.length
    findings.map(item => {
      const severity = item.finding && item.finding.severity.toLowerCase()
      switch (severity) {
      case 'low':
        low++
        break
      case 'medium':
        medium++
        break
      default:
        high++
      }
    })
    return {
      violations: [
        {count: policySet.size, violationType: msgs.get('overview.top.informations.policy.violations', locale)},
        {count: clusterSet.size, violationType: msgs.get('overview.top.informations.cluster.violations', locale)},
      ],
      findings: [
        {count: high, total: total, findingType: SECURITY_TYPES.HIGH},
        {count: medium, total: total, findingType: SECURITY_TYPES.MEDIUM},
        {count: low, total: total, findingType: SECURITY_TYPES.LOW},
      ]
    }
  }
}

const Violations = ({moduleData: {violations}, handleDrillDownClick}) => {
  return (
    <React.Fragment>
      {violations.map(({count, violationType}, idx) => {
        const onClick = () =>{
          handleDrillDownClick(violationType)
        }
        const onKeyPress = (e) =>{
          if ( e.key === 'Enter') {
            onClick()
          }
        }
        const cardClasses = classNames({
          'card-count-type': true,
          hasBorder: idx===1,
        })
        const countClasses = classNames({
          'card-count': true,
          'alert': count>0,
        })
        if (count === 0) {
          //show dash if no violations
          return (
            <div key={violationType} className={cardClasses}>
              <div className={'no-result-dash'}>
                --------
              </div>
              <div className='card-type'>
                <div className='no-result'>
                  {violationType.toUpperCase()}
                </div>
              </div>
            </div>
          )
        }
        return (
          <div key={violationType} className={cardClasses} role={'button'}
            tabIndex={0} onClick={onClick} onKeyPress={onKeyPress}>
            <div className={countClasses}>
              {count}
            </div>
            <div className='card-type'>
              <div>
                {violationType.toUpperCase()}
              </div>
            </div>
          </div>
        )
      })}
    </React.Fragment>
  )
}

Violations.propTypes = {
  handleDrillDownClick: PropTypes.func,
  moduleData: PropTypes.object,
}

const Findings = ({moduleData: {findings}, handleDrillDownClick, locale}) => {
  return (
    <React.Fragment>
      {findings.map(({count, total, findingType}) => {
        const onClick = () =>{
          handleDrillDownClick(findingType)
        }
        const onKeyPress = (e) =>{
          if ( e.key === 'Enter') {
            onClick()
          }
        }

        // set label and color
        let label, color
        switch (findingType) {
        case SECURITY_TYPES.HIGH:
          label = msgs.get('overview.recent.activity.finding.severity.high', locale)
          color = $grc_color_finding_high
          break
        case SECURITY_TYPES.MEDIUM:
          label = msgs.get('overview.recent.activity.finding.severity.medium', locale)
          color = $grc_color_finding_medium
          break
        case SECURITY_TYPES.LOW:
          label = msgs.get('overview.recent.activity.finding.severity.low', locale)
          color = $grc_color_finding_low
          break
        }

        // Generate the arc strings
        const arcs = d3.pie().sort(null)([count, Math.max(total-count, 1)])
        const pathData1 = arcGenerator(Object.assign(arcs[0], {
          innerRadius: 9,
          outerRadius: 15
        }))
        const pathData2 = arcGenerator(Object.assign(arcs[1], {
          innerRadius: 11,
          outerRadius: 13
        }))

        const countClasses = classNames({
          'card-count': true,
          'alert': count>0 && findingType===SECURITY_TYPES.HIGH,
        })
        if (count === 0) {
          //show dash if no findings
          return (
            <div key={findingType} className={'card-count-type'}>
              <div className={'no-result-dash'}>
                --------
              </div>
              <div className='card-type card-findings'>
                <div className='no-result'>
                  {label.toUpperCase()}
                </div>
              </div>
            </div>
          )
        }
        return (
          <div key={findingType} className='card-count-type' role={'button'}
            tabIndex={0} onClick={onClick} onKeyPress={onKeyPress}>
            <div className='card-count-container'>
              <svg className='card-count-pie'>
                <g>
                  <path fill={color} d={pathData1} />
                  <path fill={$grc_color_finding_ring} d={pathData2} />
                </g>
              </svg>
              <div className={countClasses}>
                {count}
              </div>
            </div>
            <div className='card-type card-findings'>
              <div>
                {label.toUpperCase()}
              </div>
            </div>
          </div>
        )
      })}
    </React.Fragment>
  )
}

Findings.propTypes = {
  handleDrillDownClick: PropTypes.func,
  locale: PropTypes.string,
  moduleData: PropTypes.object,
}

RecentActivityModule.propTypes = {
  applications: PropTypes.array,
  findings: PropTypes.array,
  handleDrillDownClick: PropTypes.func,
  policies: PropTypes.array,
  showApplications: PropTypes.bool,
  showFindings: PropTypes.bool,
  updateViewState: PropTypes.func,
  viewState: PropTypes.object,
}
