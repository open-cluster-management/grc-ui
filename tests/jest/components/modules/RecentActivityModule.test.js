/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import { BrowserRouter } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { policiesTestingDataSet1, findingsTestingDataSet1 } from './ModuleTestingData'
import RecentActivityModule from '../../../../src-web/components/modules/RecentActivityModule'

describe('RecentActivityModule view with Security Findings and applications', () => {
  jest
    .spyOn(window, 'getComputedStyle')
    .mockImplementation(() => ({display: 'block', 'padding-right': '0.625rem'}))
  const showFindings = true
  const handleDrillDownClick = jest.fn()
  const filteredPolicies = policiesTestingDataSet1
  const filteredFindings = findingsTestingDataSet1
  const viewState = {}
  const updateViewState = jest.fn()
  it('renders as expected', () => {
    const component = renderer.create(<BrowserRouter><RecentActivityModule
      showFindings={showFindings}
      showApplications={showApplications}
      policies={filteredPolicies}
      findings={filteredFindings}
      viewState={viewState}
      updateViewState={updateViewState}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component.toJSON()).toMatchSnapshot()
  })
})

describe('RecentActivityModule view without Security Findings and applications', () => {
  jest
    .spyOn(window, 'getComputedStyle')
    .mockImplementation(() => ({display: 'block', 'padding-right': '0.625rem'}))
  const showFindings = false
  const handleDrillDownClick = jest.fn()
  const filteredPolicies = policiesTestingDataSet1
  const filteredFindings = findingsTestingDataSet1
  const viewState = {}
  const updateViewState = jest.fn()
  it('renders as expected', () => {
    const component = renderer.create(<BrowserRouter><RecentActivityModule
      showFindings={showFindings}
      showApplications={showApplications}
      policies={filteredPolicies}
      findings={filteredFindings}
      viewState={viewState}
      updateViewState={updateViewState}
      handleDrillDownClick={handleDrillDownClick} /></BrowserRouter>)
    expect(component.toJSON()).toMatchSnapshot()
  })
})
