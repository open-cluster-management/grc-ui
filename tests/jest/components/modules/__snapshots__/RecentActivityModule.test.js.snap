// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentActivityModule view with Security Findings renders as expected 1`] = `
<div
  className="module-recent-activity"
>
  <div
    className="card-container-container"
  >
    <div
      className="card-title"
    >
      Recent activity
    </div>
    <div
      className="card-container"
    >
      <div
        className="card-content"
      >
        <div
          className="card-inner-content"
        >
          <div
            className="card-count-type"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="card-count alert"
            >
              3
            </div>
            <div
              className="card-type"
            >
              <div>
                POLICY VIOLATIONS
              </div>
            </div>
          </div>
          <div
            className="card-count-type hasBorder"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="card-count alert"
            >
              2
            </div>
            <div
              className="card-type"
            >
              <div>
                CLUSTER VIOLATIONS
              </div>
            </div>
          </div>
          <div
            className="card-count-type"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="card-count-container"
            >
              <svg
                className="card-count-pie"
              >
                <g>
                  <path
                    d="M9.18485099360515e-16,-15A15,15,0,1,1,-15,1.83697019872103e-15L-9,1.102182119232618e-15A9,9,0,1,0,5.51091059616309e-16,-9Z"
                    fill="#E62325"
                  />
                  <path
                    d="M-13,1.592040838891559e-15A13,13,0,0,1,-2.3880612583373385e-15,-13L-2.0206672185931327e-15,-11A11,11,0,0,0,-11,1.3471114790620886e-15Z"
                    fill="#DFE3E6"
                  />
                </g>
              </svg>
              <div
                className="card-count alert"
              >
                3
              </div>
            </div>
            <div
              className="card-type card-findings"
            >
              <div>
                HIGH SEVERITY FINDINGS
              </div>
            </div>
          </div>
          <div
            className="card-count-type"
          >
            <div
              className="no-result-dash"
            >
              --------
            </div>
            <div
              className="card-type card-findings"
            >
              <div
                className="no-result"
              >
                MEDIUM SEVERITY FINDINGS
              </div>
            </div>
          </div>
          <div
            className="card-count-type"
          >
            <div
              className="no-result-dash"
            >
              --------
            </div>
            <div
              className="card-type card-findings"
            >
              <div
                className="no-result"
              >
                LOW SEVERITY FINDINGS
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="violation-container"
    >
      <div
        className="module-top-information"
      >
        <div
          className="header-container"
        >
          <div
            className="header-title"
          >
            Top violations
          </div>
          <div
            className="header-tab"
          >
            <nav
              aria-label="Top violations undefined"
              className="bx--tabs"
              role="navigation"
              selected={0}
            >
              <div
                aria-label="listbox"
                className="bx--tabs-trigger"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="listbox"
                tabIndex={0}
              >
                <a
                  className="bx--tabs-trigger-text"
                  href="#"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Policies
                </a>
                <svg
                  alt="show menu options"
                  aria-label="show menu options"
                  className={undefined}
                  fill={undefined}
                  fillRule="evenodd"
                  height="5"
                  name={undefined}
                  role="img"
                  style={undefined}
                  viewBox="0 0 10 5"
                  width="10"
                >
                  <title>
                    show menu options
                  </title>
                  <path
                    d="M0 0l5 4.998L10 0z"
                  />
                </svg>
              </div>
              <ul
                className="bx--tabs__nav bx--tabs__nav--hidden"
                role="tablist"
              >
                <li
                  className="bx--tabs__nav-item bx--tabs__nav-item--selected"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={true}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={true}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={0}
                  >
                    Policies
                  </a>
                </li>
                <li
                  className="bx--tabs__nav-item"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={false}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={false}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={-1}
                  >
                    Clusters
                  </a>
                </li>
                <li
                  className="bx--tabs__nav-item"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={false}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={false}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={-1}
                  >
                    Applications
                  </a>
                </li>
              </ul>
            </nav>
            <div
              aria-hidden={false}
              className={undefined}
              hidden={false}
              selected={true}
            />
            <div
              aria-hidden={true}
              className={undefined}
              hidden={true}
              selected={false}
            />
            <div
              aria-hidden={true}
              className={undefined}
              hidden={true}
              selected={false}
            />
          </div>
        </div>
        <div>
          <div
            className="information-card-container"
          >
            <div>
              <div
                className="information-card-container single-strip"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              >
                <div
                  className="information-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-informations"
                    >
                      <div
                        className="card-information-count"
                      >
                        2
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name name-clickable"
                      >
                        compliance-all
                      </div>
                      <div
                        className="card-description description-clickable"
                      >
                        <div
                          className="LinesEllipsis  "
                        >
                          cluster1, clusterhub
                          <wbr />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                className="information-card-container single-strip"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              >
                <div
                  className="information-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-informations"
                    >
                      <div
                        className="card-information-count"
                      >
                        2
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name name-clickable"
                      >
                        policy-pod
                      </div>
                      <div
                        className="card-description description-clickable"
                      >
                        <div
                          className="LinesEllipsis  "
                        >
                          cluster1, clusterhub
                          <wbr />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                className="information-card-container single-strip"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              >
                <div
                  className="information-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-informations"
                    >
                      <div
                        className="card-information-count"
                      >
                        1
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name name-clickable"
                      >
                        policy-ma
                      </div>
                      <div
                        className="card-description description-clickable"
                      >
                        <div
                          className="LinesEllipsis  "
                        >
                          cluster1
                          <wbr />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        className="module-top-information"
      >
        <div
          className="header-container"
        >
          <div
            className="header-title"
          >
            Top security findings
          </div>
          <div
            className="header-tab"
          >
            <nav
              aria-label="Top security findings undefined"
              className="bx--tabs"
              role="navigation"
              selected={0}
            >
              <div
                aria-label="listbox"
                className="bx--tabs-trigger"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="listbox"
                tabIndex={0}
              >
                <a
                  className="bx--tabs-trigger-text"
                  href="#"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Security findings
                </a>
                <svg
                  alt="show menu options"
                  aria-label="show menu options"
                  className={undefined}
                  fill={undefined}
                  fillRule="evenodd"
                  height="5"
                  name={undefined}
                  role="img"
                  style={undefined}
                  viewBox="0 0 10 5"
                  width="10"
                >
                  <title>
                    show menu options
                  </title>
                  <path
                    d="M0 0l5 4.998L10 0z"
                  />
                </svg>
              </div>
              <ul
                className="bx--tabs__nav bx--tabs__nav--hidden"
                role="tablist"
              >
                <li
                  className="bx--tabs__nav-item bx--tabs__nav-item--selected"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={true}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={true}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={0}
                  >
                    Security findings
                  </a>
                </li>
                <li
                  className="bx--tabs__nav-item"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={false}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={false}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={-1}
                  >
                    Clusters
                  </a>
                </li>
              </ul>
            </nav>
            <div
              aria-hidden={false}
              className={undefined}
              hidden={false}
              selected={true}
            />
            <div
              aria-hidden={true}
              className={undefined}
              hidden={true}
              selected={false}
            />
          </div>
        </div>
        <div>
          <div
            className="information-card-container"
          >
            <div>
              <div
                className="information-card-container single-strip"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              >
                <div
                  className="information-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-informations"
                    >
                      <div
                        className="card-information-count"
                      >
                        3
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name name-clickable"
                      >
                        Policy that is not compliant
                      </div>
                      <div
                        className="card-description description-clickable"
                      >
                        <div
                          className="LinesEllipsis  "
                        >
                          mycluster_name, cluster1, clusterhub
                          <wbr />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                className="information-card-container single-strip empty-strip"
              >
                <div
                  className="information-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-informations empty"
                    >
                      <img
                        alt="No resource"
                        className="empty-icon"
                        src="undefined/graphics/no-other-violations.svg"
                      />
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name name-empty"
                      >
                        No other security findings
                      </div>
                      <div
                        className="card-description description-empty"
                      >
                        <div
                          className="LinesEllipsis  "
                        >
                          We will continue to monitor and display any security findings so you can easily find them here.
                          <wbr />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`RecentActivityModule view without Security Findings renders as expected 1`] = `
<div
  className="module-recent-activity"
>
  <div
    className="card-container-container"
  >
    <div
      className="card-title"
    >
      Recent activity
    </div>
    <div
      className="card-container"
    >
      <div
        className="card-content"
      >
        <div
          className="card-inner-content"
        >
          <div
            className="card-count-type"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="card-count alert"
            >
              3
            </div>
            <div
              className="card-type"
            >
              <div>
                POLICY VIOLATIONS
              </div>
            </div>
          </div>
          <div
            className="card-count-type hasBorder"
            onClick={[Function]}
            onKeyPress={[Function]}
            role="button"
            tabIndex={0}
          >
            <div
              className="card-count alert"
            >
              2
            </div>
            <div
              className="card-type"
            >
              <div>
                CLUSTER VIOLATIONS
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      className="violation-container"
    >
      <div
        className="module-top-information"
      >
        <div
          className="header-container"
        >
          <div
            className="header-title"
          >
            Top violations
          </div>
          <div
            className="header-tab"
          >
            <nav
              aria-label="Top violations undefined"
              className="bx--tabs"
              role="navigation"
              selected={0}
            >
              <div
                aria-label="listbox"
                className="bx--tabs-trigger"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="listbox"
                tabIndex={0}
              >
                <a
                  className="bx--tabs-trigger-text"
                  href="#"
                  onClick={[Function]}
                  tabIndex={-1}
                >
                  Policies
                </a>
                <svg
                  alt="show menu options"
                  aria-label="show menu options"
                  className={undefined}
                  fill={undefined}
                  fillRule="evenodd"
                  height="5"
                  name={undefined}
                  role="img"
                  style={undefined}
                  viewBox="0 0 10 5"
                  width="10"
                >
                  <title>
                    show menu options
                  </title>
                  <path
                    d="M0 0l5 4.998L10 0z"
                  />
                </svg>
              </div>
              <ul
                className="bx--tabs__nav bx--tabs__nav--hidden"
                role="tablist"
              >
                <li
                  className="bx--tabs__nav-item bx--tabs__nav-item--selected"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={true}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={true}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={0}
                  >
                    Policies
                  </a>
                </li>
                <li
                  className="bx--tabs__nav-item"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={false}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={false}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={-1}
                  >
                    Clusters
                  </a>
                </li>
                <li
                  className="bx--tabs__nav-item"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="presentation"
                  selected={false}
                  tabIndex={-1}
                >
                  <a
                    aria-selected={false}
                    className="bx--tabs__nav-link"
                    href="#"
                    role="tab"
                    tabIndex={-1}
                  >
                    Applications
                  </a>
                </li>
              </ul>
            </nav>
            <div
              aria-hidden={false}
              className={undefined}
              hidden={false}
              selected={true}
            />
            <div
              aria-hidden={true}
              className={undefined}
              hidden={true}
              selected={false}
            />
            <div
              aria-hidden={true}
              className={undefined}
              hidden={true}
              selected={false}
            />
          </div>
        </div>
        <div>
          <div
            className="information-card-container"
          >
            <div>
              <div
                className="information-card-container single-strip"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              >
                <div
                  className="information-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-informations"
                    >
                      <div
                        className="card-information-count"
                      >
                        2
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name name-clickable"
                      >
                        compliance-all
                      </div>
                      <div
                        className="card-description description-clickable"
                      >
                        <div
                          className="LinesEllipsis  "
                        >
                          cluster1, clusterhub
                          <wbr />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                className="information-card-container single-strip"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              >
                <div
                  className="information-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-informations"
                    >
                      <div
                        className="card-information-count"
                      >
                        2
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name name-clickable"
                      >
                        policy-pod
                      </div>
                      <div
                        className="card-description description-clickable"
                      >
                        <div
                          className="LinesEllipsis  "
                        >
                          cluster1, clusterhub
                          <wbr />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div>
              <div
                className="information-card-container single-strip"
                onClick={[Function]}
                onKeyPress={[Function]}
                role="button"
                tabIndex={0}
              >
                <div
                  className="information-card-content"
                >
                  <div
                    className="card-inner-content"
                  >
                    <div
                      className="card-informations"
                    >
                      <div
                        className="card-information-count"
                      >
                        1
                      </div>
                    </div>
                    <div
                      className="card-name-description"
                    >
                      <div
                        className="card-name name-clickable"
                      >
                        policy-ma
                      </div>
                      <div
                        className="card-description description-clickable"
                      >
                        <div
                          className="LinesEllipsis  "
                        >
                          cluster1
                          <wbr />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
